blog/_src/posts/2019-05-14-scoping-in-r.md

```diff
@@ -0,0 +1,329 @@
+new binding to that environment, where `x` has the value `4`. Finally, on line
+5, we perform a variable lookup in environment `e`.
+
+#### Deleting bindings
+
+```r
+x <- 1
+f <- function(a) {
+  rm("x", envir = parent.env(environment()))
+  x + a
+}
+g <- function() {
+  x <- 2
+  f(0)
+}
+g() # Error in f(0) : object 'x' not found
+```
+
+On line 3, we delete the binding `x` from the given environment,
+`parent.env(environment())`. This refers to the parent (or enclosing)
+environment of the current environment of `f`, i.e. the environment where `f`
+was defined, which is the top-level environment.
+
+
+### What does "scoping in R is dynamic" mean?
+
+In _Optimizing R Language Execution via Aggressive Speculation_ (DLS 2016)
+([doi][swhj2016-doi]), Stadler, Welc, Humer, and Jordan write:
+
+> [...] variable scoping in R is dynamic and can be modified at the language
+> level [...]
+
+[swhj2016-doi]: https://doi.org/10.1145/2989225.2989236
+
+Our previous examples should demonstrate that scoping in R "can be modified at
+the language level."
+
+It is also true that "variable scoping in R is dynamic," in the sense that
+scoping is determined at run time. However, this is more general than the
+definition of _dynamic scoping_ we used in this blog post.
+
+Similarly, I've seen the term _lexical scoping_ used with a less specific but
+stronger definition---that all variable bindings can be determined at compile
+time. Under these more general definitions, R would be dynamically scoped
+rather than lexically scoped.
+
+I prefer the more specific definitions of lexical and dynamic scoping; I think
+they're more useful. Either way, it's important to be clear and make sure
+your audience knows what definitions you're using!
+
+---
+
+## Extra: another taxonomy for scoping rules
+
+In _[Lexical Scope and Statistical Computation][gi2000]_ (2000)
+([doi][gi2000-doi]), Gentleman and Ihaka, the creators of R, define four kinds
+of scoping rules for the purpose of their discussion:
+
+[gi2000]: https://www.stat.auckland.ac.nz/~ihaka/downloads/lexical.pdf
+[gi2000-doi]: https://doi.org/10.1080/10618600.2000.10474895
+
+  - *trivial*: no free variables allowed
+  - *dynamic*: as we defined it; free variables are evaluated in their dynamic
+     context
+  - *static*: values of free variables are defined by global variables, e.g.
+    scoping in C
+  - *lexical*: as we defined it; free variables are evaluated in their lexical
+     context
+
+The paper is written for a statistics audience, and the authors explain how
+lexical scoping in R is useful. They provide motivating examples from
+statistical computing, illustrating how lexical scoping in R is an advantage
+over the scoping rules of S, the language R was based on.
+
+My colleague, Jan Ječmen, came up with examples of the four scoping rules.
+(I edited them slightly for this blog post.)
+
+
+### Trivial scoping
+
+```r
+x <- 0
+makeTrivial <- function() {
```

Comment:
It'd be super nice if you could decouple the actual code (`\a -> a + x`) from the environment management. Could all `make*` accept a lambda with free variables? Would it turn the implementation of `make*` into a much harder task?

Reply:
There's probably a way in R to do this, with `quote` (which returns an unevaluated expression) and `eval`, but that would be uglier.

I might just remove this section entirely...

Reply:
Wait, I keep hearing that R is lazy by default for the very purpose of allowing kind of `quote` behavior *by default*! If I got this right, the trick I'm thinking of should be very cheap on the user side — it might complicate a bit the implementation, though. As I'm no R expert, maybe I'm wrong, but I'd summon the Gods of R to consult! @aviralg @janjecmen @o-.

I get you're reluctant about this section. I say two things. First, I found its idea exciting, so I'll miss it if it's gone. Second, it's a bit technical, so I didn't read it really closely, so you're right that it's a bit different from the first part. But I see no harm in the second: this is not a paper, and there might be people on the internet who admire it.

Reply:
I can do something like:
```r
x <- 0
makeStatic <- function(fun) {
    x <- 1
    res <- eval(substitute(fun))
    environment(res) <- globalenv()
    res
}

f <- makeStatic(function(a) x+a)
g <- function() {
  x <- 2
  f(0)
}
g() # 0
```

But I haven't figured out how to make this trick work for dynamic scoping...

Reply:
Wow this is a hack:

```r
makeDynamic <- function(fun) {
    x <- 1
    function(...) {
        res <- eval(substitute(fun))
        environment(res) <- new.env(parent = parent.frame())
        res(...)
    }
}

f <- makeDynamic(function(a) x+a)
g <- function() {
  x <- 2
  f(0)
}
g() # 2
```

Reply:
I wanted a non-trivial function that also uses a bound variable.